File: valid-palindrome/rivkode.py
import re


# Time Complexity O(n)
# - Both normal preprocessing and two-pointer comparisons have time complexity proportional to the length of the input string.
# Space Complexity O(n)
# - The space of the new string joined_string preprocessed from the input string is proportional to the length of the input string.

class Solution:
    def isPalindrome(self, s: str) -> bool:
        # Pre-processing using regular expression
        joined_string = re.sub(r"[^a-zA-Z]", "", s.lower())
        str_length = len(joined_string)


        # for loop n/2, two pointer for verify same or not
        for i in range(str_length // 2):
            end = str_length - i - 1
            if not self.check_palindrome(i, end, joined_string):
                return False

        return True

    def check_palindrome(self, i, end, joined_string) -> bool:
        left = joined_string[i]
        right = joined_string[end]

        if left == right:
            return True
        else:
            return False

if __name__ == "__main__":
    solution = Solution()

    # test case
    test_string = [
        "A man, a plan, a canal: Panama",  # True (회문)
        "race a car",  # False (회문 아님)
        " ",  # True (빈 문자열도 회문으로 간주)
        "abba",  # True (회문)
        "abcba",  # True (회문)
    ]

    for index, test in enumerate(test_string):
        print(f"start {index} test")
        print(f"input : {test}")
        print(f"Is valid palindrome ? {solution.isPalindrome(test)}\n")

File: valid-palindrome/rivkode.java
/*
1. 문제 이해
입력받은 글자에 대해 뒤집어도 동일한 글자인지 아닌지를 판단

2. 예외 케이스
빈 문자열, 글자가 아닌 다른 내용일 경우 제거

3. 알고리즘
단순 루프

4. 구현

글자를 입력받고 글자가 아닌 내용에 대해서는 필터링한다
전체 길이를 구하고 2포인터로 왼쪽, 오른쪽을 서로 줄여가면서
같거나 교차할때 종료

다른 부분이 있으면 false 반환, 모두 통과하면 true 반환

빈 값일경우 바로 true


*/

import java.util.*;

class Solution {
    public boolean isPalindrome(String s) {
        String result = s.replaceAll("[^a-zA-Z0-9]", "");
        result = result.toLowerCase();

        int len = result.length();
        int left = 0;
        int right = len - 1;

        while (left < right) {
            char leftS = result.charAt(left);
            char rightS = result.charAt(right);

            if (leftS != rightS) {
                return false;
            }

            left += 1;
            right -= 1;
        }

        return true;
    }
}

